source/extensions/filters/network/mysql_proxy/conn_pool_impl.h

```diff
@@ -0,0 +1,187 @@
+    void onProtocolError() override;
+    void onNewMessage(MySQLSession::State) override{};
+    void onServerGreeting(ServerGreeting&) override;
+    void onClientLogin(ClientLogin&) override;
+    void onClientLoginResponse(ClientLoginResponse&) override;
+    void onClientSwitchResponse(ClientSwitchResponse&) override;
+    void onMoreClientLoginResponse(ClientLoginResponse&) override;
+    void onCommand(Command&) override;
+    void onCommandResponse(CommandResponse&) override;
+
+    // ConnectionPool::UpstreamCallbacks
+    void onUpstreamData(Buffer::Instance&, bool) override;
+    void onEvent(Network::ConnectionEvent event) override;
+    void onAboveWriteBufferHighWatermark() override {}
+    void onBelowWriteBufferLowWatermark() override {}
+
+    // Tcp::ConnectionPool::Callbacks
+    void onPoolFailure(Tcp::ConnectionPool::PoolFailureReason reason,
+                       Upstream::HostDescriptionConstSharedPtr host) override;
+    void onPoolReady(Tcp::ConnectionPool::ConnectionDataPtr&& conn,
+                     Upstream::HostDescriptionConstSharedPtr host) override;
+
+    void writeUpstream(MySQLCodec&, uint8_t);
+    void onFailure(MySQLPoolFailureReason reason);
+    void onPassConnectionPhase();
+
+    ThreadLocalClientPool& parent_;
+    ClientState state_{ClientState::Uinit};
+    std::vector<uint8_t> old_auth_data_;
+    ClientWrapperSharedPtr client_wrapper_;
+  };
+
+  struct PendingRequest : public LinkedObject<PendingRequest>, public Cancellable {
+  public:
+    PendingRequest(ThreadLocalClientPool& parent, ClientPoolCallBack&);
+    void cancel() override;
+    ThreadLocalClientPool& parent_;
+    ClientPoolCallBack& callbacks_;
+  };
+
+  using ThreadLocalActiveClientPtr = std::unique_ptr<ThreadLocalActiveClient>;
+  using PendingRequestPtr = std::unique_ptr<PendingRequest>;
+
+  struct ThreadLocalClientPool : public ThreadLocal::ThreadLocalObject {
+    ThreadLocalClientPool(std::shared_ptr<InstanceImpl> parent, Upstream::ClusterManager* cm,
+                          Event::Dispatcher& dispatcher, const ConnectionPoolSettings& config,
+                          DecoderFactory& decoder_factory, const std::string& auth_username,
+                          const std::string& auth_password);
+    ~ThreadLocalClientPool() override;
+    Cancellable* newMySQLClient(ClientPoolCallBack&);
+    void processIdleClient(ThreadLocalActiveClient&, bool new_client);
+    void createNewClient();
+    void initStartClients();
+    void removeClient(ThreadLocalActiveClientPtr&& client);
+
+    std::weak_ptr<InstanceImpl> parent_;
+    Event::Dispatcher& dispatcher_;
+    Tcp::ConnectionPool::Instance* conn_pool_{nullptr};
+    std::list<PendingRequestPtr> pending_requests_;
+    DecoderFactory& decoder_factory_;
+    ConnectionPoolSettings config_;
+    std::list<ThreadLocalActiveClientPtr> pending_clients_;
+    std::list<ThreadLocalActiveClientPtr> active_clients_;
+    std::list<ThreadLocalActiveClientPtr> busy_clients_;
```

Comment:
connection reuse will be awesome! That said, this looks a lot like the legacy TCP connection pool and I don't think we want to reinvent this again, especially as the base pool supports connection reuse.

Can we please inherit from souce/common/conn_pool/conn_pool_base.* and refactor/extend where needed to avoid all the duplicate pending/active/busy work and reuse stats etc?

Reply:
Yes, that's better.